File: sdk/python/soluna/relay.py
"""
Low-level BSD socket connection to the Soluna relay server.
Uses UDP sockets, matching the Swift/C++ implementation.
"""

from __future__ import annotations

import socket
import struct
import threading
import time
from typing import Callable

from .parser import OSTConstants


class RelayConnection:
    """UDP socket connection to the Soluna relay."""

    def __init__(
        self,
        channel: str,
        host: str = OSTConstants.DEFAULT_HOST,
        port: int = OSTConstants.DEFAULT_PORT,
        device_name: str = "SolunaSDK-Python",
    ) -> None:
        self.channel = channel
        self.host = host
        self.port = port
        self.device_name = device_name

        self._sock: socket.socket | None = None
        self._running = False
        self._recv_thread: threading.Thread | None = None
        self._heartbeat_thread: threading.Thread | None = None
        self._addr: tuple[str, int] | None = None

        self.on_packet: Callable[[bytes], None] | None = None
        self.on_control_message: Callable[[str], None] | None = None

    def connect(self) -> bool:
        """Open the UDP socket, send HELLO/JOIN, and start the receive loop.
        Returns False if DNS resolution or socket creation fails.
        """
        if self._running:
            return True

        try:
            # DNS resolve
            info = socket.getaddrinfo(self.host, self.port, socket.AF_INET, socket.SOCK_DGRAM)
            if not info:
                return False
            self._addr = info[0][4]
        except socket.gaierror:
            return False

        # Create UDP socket
        self._sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        self._sock.settimeout(1.0)

        # Send HELLO x3 (100ms apart)
        for i in range(3):
            self._send_message("HELLO\n")
            if i < 2:
                time.sleep(0.1)

        # JOIN
        self._send_message(f"JOIN:{self.channel}::{self.device_name}\n")

        self._running = True

        # Start receive thread
        self._recv_thread = threading.Thread(target=self._recv_loop, daemon=True)
        self._recv_thread.start()

        # Start heartbeat thread
        self._heartbeat_thread = threading.Thread(target=self._heartbeat_loop, daemon=True)
        self._heartbeat_thread.start()

        return True

    def disconnect(self) -> None:
        """Close the connection."""
        self._running = False

        if self._sock:
            self._sock.close()
            self._sock = None

        if self._recv_thread:
            self._recv_thread.join(timeout=2.0)
            self._recv_thread = None

        if self._heartbeat_thread:
            self._heartbeat_thread.join(timeout=2.0)
            self._heartbeat_thread = None

    def _send_message(self, message: str) -> None:
        if self._sock and self._addr:
            try:
                self._sock.sendto(message.encode("utf-8"), self._addr)
            except OSError:
                pass

    def _recv_loop(self) -> None:
        while self._running and self._sock:
            try:
                data, _ = self._sock.recvfrom(OSTConstants.RECV_BUFFER_SIZE)
            except socket.timeout:
                continue
            except OSError:
                break

            if len(data) < OSTConstants.RTP_HEADER_SIZE:
                continue

            # RTP/OSTP audio packet: (byte[0] & 0xC0) == 0x80
            if (data[0] & 0xC0) == 0x80:
                if self.on_packet:
                    self.on_packet(data)
            else:
                # Text control message
                if self.on_control_message:
                    try:
                        msg = data.decode("utf-8")
                        self.on_control_message(msg)
                    except UnicodeDecodeError:
                        pass

    def _heartbeat_loop(self) -> None:
        while self._running:
            time.sleep(OSTConstants.HEARTBEAT_INTERVAL)
            if not self._running:
                break
            self._send_message("HELLO\n")
            self._send_message(f"JOIN:{self.channel}::{self.device_name}\n")

File: sdk/rust/src/relay.rs
//! Low-level UDP socket connection to the Soluna relay server.
//! Uses std::net::UdpSocket, matching the Swift/C++ implementation.

use std::io;
use std::net::{ToSocketAddrs, UdpSocket};
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::thread;
use std::time::Duration;

use crate::parser::constants;

/// UDP socket connection to the Soluna relay.
pub struct RelayConnection {
    channel: String,
    host: String,
    port: u16,
    device_name: String,
    socket: Option<UdpSocket>,
    running: Arc<AtomicBool>,
    recv_handle: Option<thread::JoinHandle<()>>,
    heartbeat_handle: Option<thread::JoinHandle<()>>,
}

impl RelayConnection {
    pub fn new(channel: &str, host: &str, port: u16, device_name: &str) -> Self {
        Self {
            channel: channel.to_string(),
            host: host.to_string(),
            port,
            device_name: device_name.to_string(),
            socket: None,
            running: Arc::new(AtomicBool::new(false)),
            recv_handle: None,
            heartbeat_handle: None,
        }
    }

    /// Open the UDP socket, send HELLO/JOIN, and start the receive loop.
    /// Returns an error if DNS resolution or socket creation fails.
    ///
    /// `on_packet` is called on the recv thread when an audio packet arrives.
    /// `on_control` is called on the recv thread for text control messages.
    pub fn connect<F, G>(
        &mut self,
        on_packet: F,
        on_control: G,
    ) -> io::Result<()>
    where
        F: Fn(&[u8]) + Send + 'static,
        G: Fn(&str) + Send + 'static,
    {
        if self.running.load(Ordering::SeqCst) {
            return Ok(());
        }

        // DNS resolve
        let addr_str = format!("{}:{}", self.host, self.port);
        let addr = addr_str
            .to_socket_addrs()?
            .find(|a| a.is_ipv4())
            .ok_or_else(|| io::Error::new(io::ErrorKind::NotFound, "DNS resolution failed"))?;

        // Create UDP socket
        let socket = UdpSocket::bind("0.0.0.0:0")?;
        socket.set_read_timeout(Some(Duration::from_secs(1)))?;
        socket.connect(addr)?;

        // Send HELLO x3 (100ms apart)
        for i in 0..3 {
            let _ = socket.send(b"HELLO\n");
            if i < 2 {
                thread::sleep(Duration::from_millis(100));
            }
        }

        // JOIN
        let join_msg = format!("JOIN:{}::{}\n", self.channel, self.device_name);
        let _ = socket.send(join_msg.as_bytes());

        self.running.store(true, Ordering::SeqCst);

        // Clone for threads
        let recv_socket = socket.try_clone()?;
        let running_recv = self.running.clone();

        // Receive thread
        self.recv_handle = Some(thread::spawn(move || {
            let mut buf = vec![0u8; constants::RECV_BUFFER_SIZE];
            while running_recv.load(Ordering::SeqCst) {
                match recv_socket.recv(&mut buf) {
                    Ok(n) if n >= constants::RTP_HEADER_SIZE => {
                        if (buf[0] & 0xC0) == 0x80 {
                            on_packet(&buf[..n]);
                        } else if let Ok(msg) = std::str::from_utf8(&buf[..n]) {
                            on_control(msg);
                        }
                    }
                    _ => continue,
                }
            }
        }));

        // Heartbeat thread
        let heartbeat_socket = socket.try_clone()?;
        let running_hb = self.running.clone();
        let channel = self.channel.clone();
        let device_name = self.device_name.clone();

        self.heartbeat_handle = Some(thread::spawn(move || {
            while running_hb.load(Ordering::SeqCst) {
                thread::sleep(Duration::from_secs(constants::HEARTBEAT_INTERVAL_SECS));
                if !running_hb.load(Ordering::SeqCst) {
                    break;
                }
                let _ = heartbeat_socket.send(b"HELLO\n");
                let join_msg = format!("JOIN:{}::{}\n", channel, device_name);
                let _ = heartbeat_socket.send(join_msg.as_bytes());
            }
        }));

        self.socket = Some(socket);
        Ok(())
    }

    /// Close the connection.
    pub fn disconnect(&mut self) {
        self.running.store(false, Ordering::SeqCst);

        // Drop the socket to unblock recv
        self.socket = None;

        if let Some(h) = self.recv_handle.take() {
            let _ = h.join();
        }
        if let Some(h) = self.heartbeat_handle.take() {
            let _ = h.join();
        }
    }

    pub fn is_connected(&self) -> bool {
        self.running.load(Ordering::SeqCst)
    }
}

impl Drop for RelayConnection {
    fn drop(&mut self) {
        self.disconnect();
    }
}
